### ISSUE TITLE
Irrelevant sampling affects MCMC marginal

### ISSUE THREAD

**Issue by mhtess** [probmods/webppl#495]
This model

```
var n = 15
var model = function() {
  var p = uniform(0,1)    // sample a coin weight from a uniform distribution
  var totallyIrrelevantGarbage = binomial(p, n)
  return p
}

Infer({samples: 10000, burn: 5000, verbose:true}, model)
```

does not return uniform. Changing `binomial` to `gaussian` returns uniform, appropriately.

Some notes [as an issue at RWebPPL](https://github.com/mhtess/rwebppl/issues/20)

Thanks to @hawkrobe for discovering this

**Comment by mhtess** [comment]
A little more context. We get this behavior on using rwebppl (which uses the most recent install from npm)  and webppl.org

This is what we ran on webppl.org
```
var n = 15
var model = function() {
  var p = uniform(0,1)    // sample a coin weight from a uniform distribution
  var totallyIrrelevantGarbage = binomial(p, n)
  return p
}
var rs = MCMC(model, {samples: 10000, burn: 5000, verbose:true})
vizPrint(rs)
```

**Comment by longouyang** [comment]
I did some more testing and found that:
1. I get this weird behavior if the irrelevant distribution is Bernoulli, Binomial, Gamma, Dirichlet, Poisson, or Beta, but *not* Uniform or Gaussian.
2. I don't seem to get this issue at all with rejection sampling.

**Comment by null-a** [comment]
I think the issue @mhtess reports is a manifestation of the bugs that were in the binomial distribution, and that it will be fixed by af0fa9b8. This came after `v0.7`, so this is consistent with seeing a problem with the latest npm release and webppl.org.

>  I get this weird behavior if the irrelevant distribution is Bernoulli, Binomial, Gamma, Dirichlet, Poisson, or Beta, but not Uniform or Gaussian.

I'm not sure what's going on here. I've only been able to reproduce the wonky histogram using the binomial. (And a version of webppl prior to the commit i mention above.)

> The issue seems related to MCMC internals -- I get an acceptance ratio of 1 if I comment out the irrelevant line but 0.83 if I leave it in.

I think that's expected. The acceptance ratio would be 1 if the two variables were independent, but when they are dependent proposals to the first variable can be rejected.

**Comment by longouyang** [comment]
I just tested with the latest webppl, `v0.7.0-8941a56`.

Running without irrelevant sample:
```js
util.seedRNG(1)
var model = function() {
  var p = uniform(0,1);
  //var totallyIrrelevantGarbage = flip(0.99 * p)
  return p
}
viz.auto(MCMC(model, {samples: 50000, burn: 5000}))
```

Gives a uniform distribution:
![0af2af](https://cloud.githubusercontent.com/assets/14802/16051574/2dbe308c-3215-11e6-90fd-cd2bd83c7b19.png)

Using the Bernoulli irrelevant distribution:

```js
util.seedRNG(1)
var model = function() {
  var p = uniform(0,1);
  var totallyIrrelevantGarbage = flip(0.99 * p)
  return p
}
viz.auto(MCMC(model, {samples: 50000, burn: 5000}))
```

Gives a non-uniform distribution:

![b4e40f](https://cloud.githubusercontent.com/assets/14802/16051611/4ba13108-3215-11e6-8b27-fc9e4e47a0f5.png)

This is a bit easier to see if query on the distance to the center, `Math.abs(p - 0.5)`:

![119fe2](https://cloud.githubusercontent.com/assets/14802/16051664/87323a00-3215-11e6-9764-b842b0ea74e8.png)

So I think that the problem is not fixed by the recent binomial improvements.

**Comment by stuhlmueller** [comment]
Are you sure this isn't just about MCMC mixing times? If you introduce the additional variable, MCMC will take longer to converge if there is strong dependence between the two variables.

**Comment by longouyang** [comment]
Ah, this does appear to be about the mixing times -- running with 10^6 samples resulted in a uniform distribution (FWIW, I can now get the "bug" for Gaussian by scaling the params, e.g., `gaussian(p * 10, 1)`).

It was kind of surprising that I had to run this many samples, though. We should probably write down some notes on inference pitfalls (for, e.g., the PPAML course and dippl) and include this there.

**Comment by null-a** [comment]
The original problem created a very distinctive wonky histogram that I felt quite sure was fixed by the commit I mentioned.

I didn't used webppl-viz when I was looking at this. But playing around just now with the version [here](http://probmods.github.io/webppl-viz/) I noticed that the result looks more uniform if you switch from `viz.auto` to `viz.hist`?

![thing](https://cloud.githubusercontent.com/assets/9109012/16053192/9d86b5f2-325f-11e6-8be7-f2c94e94bef2.png)

(That's using the same random seed as in your example.)

Is there any chance the kernel density thing is assuming all elements of the support have the same probability?
